src/core/linux/SDL_dbus.c

```diff
@@ -359,8 +377,68 @@ SDL_DBus_ScreensaverTickle(void)
 SDL_bool
 SDL_DBus_ScreensaverInhibit(SDL_bool inhibit)
 {
-    if ( (inhibit && (screensaver_cookie != 0)) || (!inhibit && (screensaver_cookie == 0)) ) {
+    const char *default_inhibit_reason = "Playing a game";
+
+    if ( (inhibit && (screensaver_cookie != 0 || inhibit_handle != NULL))
+        || (!inhibit && (screensaver_cookie == 0 && inhibit_handle == NULL)) ) {
         return SDL_TRUE;
+    }
+
+    /* Check if we are running inside a Flatpak or Snap container.
+     * For Snap, we check multiple variables because they might be set for unrelated
+     * reasons. This is the same thing WebKitGTK does. */
+    if (access("/.flatpak-info", F_OK) == 0
+        || (SDL_getenv("SNAP") != NULL && SDL_getenv("SNAP_NAME") != NULL && SDL_getenv("SNAP_REVISION") != NULL)) {
+
+        const char *node = "org.freedesktop.portal.Desktop";
+        const char *path = "/org/freedesktop/portal/desktop";
+        const char *interface = "org.freedesktop.portal.Inhibit";
+        const char *window = ""; /* As a future improvement we could gather the X11 XID or Wayland surface identifier */
+        static unsigned int INHIBIT_IDLE = 8; /* Taken from the portal API reference */
+        DBusMessageIter iterInit, iterDict, iterEntry, iterValue;
+
+        if (inhibit) {
+            DBusMessage *msg;
+            SDL_bool retval = SDL_FALSE;
+            const char *key = "reason";
+            char *reply = NULL;
+            const char *reason = SDL_GetHint(SDL_HINT_SCREENSAVER_INHIBIT_ACTIVITY_NAME);
+            if (!reason || !reason[0]) {
+                reason = default_inhibit_reason;
+            }
+
+            msg = dbus.message_new_method_call(node, path, interface, "Inhibit");
+            if (!msg) {
+                return SDL_FALSE;
+            }
+
+            dbus.message_append_args(msg, DBUS_TYPE_STRING, &window, DBUS_TYPE_UINT32, &INHIBIT_IDLE, DBUS_TYPE_INVALID);
+            dbus.message_iter_init_append(msg, &iterInit);
```

Comment:
I'm probably going to regret asking this, but: what's SDL's policy on out-of-memory? In theory most of these functions can fail (if `malloc()` or `realloc()` returns `NULL`). (Of course in reality that never happens, at least on Linux, because the kernel overcommits instead.)

There are generally two reasonable policies: the GLib/GTK policy (detect the error condition as unrecoverable and `abort()`) and the libdbus policy (do your best to unwind the error).

Reply:
I tried to move all these `iter_open` and `iter_append` to a separate function to properly check the returned bool for all of them.
Then I used the common (?) pattern of calling `iter_close_container` for all the opened containers, and within a `goto failed` I used `iter_abandon_container_if_open`.

However `dbus_message_iter_abandon_container_if_open` has been introduced only 5 years ago and SDL doesn't seem to require a particular minimum DBus version. So, should I assume that we can't rely on `abandon_container_if_open` being always available?

Reply:
You can only use `dbus_message_iter_abandon_container_if_open` if the iterator was initialized with zeroes. If it contained arbitrary uninitialized bytes from the stack, then `dbus_message_iter_abandon_container_if_open` will do the wrong thing. I added it in 2017 to make error unwinding less inconvenient.

`dbus_message_iter_abandon_container` was added in 2009. To use it correctly, you have to keep track of the state of the iterator yourself:

* never opened (uninitialized or zero-filled)? do nothing
* already closed? do nothing
* opened but not closed? `abandon` it on error

(Yes, I know this is inconvenient, and that's why I added `dbus_message_iter_abandon_container_if_open`.)

If libdbus is a sufficiently ancient version that it doesn't have `dbus_message_iter_abandon_container` either, then there is no correct thing that you can do. The least bad thing you can do is to do nothing and just let the memory leak (which is obviously bad, but this is an out-of-memory situation that doesn't happen in practice, so...)

Reply:
The reason I asked about SDL's policy is that I'm familiar with both the real-world policies for what happens when trying to allocate a small amount of memory fails: trying to do error-unwinding (as exemplified by dbus), and treating it as a fatal error (as exemplified by GLib).

I'm not a SDL maintainer, so it's not up to me which policy SDL follows, but [implementation experience from dbus](https://blog.ometer.com/2008/02/04/out-of-memory-handling-d-bus-experience/) is that correctly unwinding from OOM errors adds a significant amount of code (estimated by a dbus maintainer to be +30-40% lines of code), and testing that code will probably reveal that it's doing something wrong anyway. This is why GLib's policy on OOM is that it's generally treated as a fatal error.

Reply:
Currently I just return without using `dbus_message_iter_abandon_container` nor `dbus_message_iter_abandon_container_if_open`. I'm not sure if we are allowed to use them. Given that this will happen only in oom situations, in theory it shouldn't be an issue to leave the containers open.

If we can use `dbus_message_iter_abandon_container*` then it will be trivial to add it.

Reply:
> I'm probably going to regret asking this, but: what's SDL's policy on out-of-memory?

(I'm answering this without reading the rest of this thread, but...) unwind the error; we never terminate the process for a malloc failure in in any part of SDL.